app/providers/atmosphere/optimization_strategy/default.rb

```diff
@@ -35,17 +35,39 @@ def can_scale_manually?
       def vmt_candidates_for(appliance)
         vmts = VirtualMachineTemplate.where(
           appliance_type: appliance.appliance_type,
-          state: 'active',
+          state: 'active'
         )
-        vmts.select do |vmt|
-          (vmt.tenants & appliance.tenants.active.funded_by(appliance.fund)).present?
+        if appliance.tenants.present?
+          vmts = restrict_by_user_requirements(vmts, appliance)
         end
+        vmts = restrict_by_tenant_availability(vmts, appliance)
+        vmts
       end
 
       private
 
       attr_reader :appliance
 
+      def restrict_by_user_requirements(vmts, appliance)
+        # If the user requests that the appliance be bound to a specific set of tenants,
+        # the optimizer should honor this selection.
+
+        # Workaround
+        funded_active_tenants = appliance.tenants.funded_by(appliance.fund).reject{|t| t.active == false}
+
+        vmts.select do |vmt|
+          (vmt.tenants & funded_active_tenants).present?
```

Comment:
OK, I take it we cannot keep the existing code.

(...1 hour later...)

I think this *probably* does what you want:

<pre>
Atmosphere::VirtualMachineTemplate.joins(:tenants).
where(Atmosphere::VirtualMachineTemplate.arel_table[:id].in(vmts.pluck(:id)).
and(Atmosphere::Tenant.arel_table[:id].in(funded_active_tenants.pluck(:id))))
</pre>

Of course if you know of a better way to write this query (as opposed to "something like this query" ;)) then go ahead and tell me.

God I hate Arel... :angry: